### PR TITLE
Load api

### PR DESCRIPTION
**Pull request by eracah**
# What does this PR do?

# What issue(s) does this change relate to?

# Before submitting
- [ ] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [ ] Did you update any related docs and document your change?
- [ ] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [ ] Did you run the tests locally to make sure they pass?
- [ ] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

File: composer/checkpoint/load.py [hunk 1]
```diff
@@ -0,0 +1,519 @@
...
+    load_model_and_optim_together = (load_options.load_optimizer and load_options.load_model
+                                    ) and (load_options.sharded_checkpoint or load_options.shard_as_needed_during_load)
+    if load_model_and_optim_together:
+        load_model_checkpoint(model, load_path=load_path, optimizer=optim, load_options=load_options)
+
+    else:
+        load_model_checkpoint(model, load_path=load_path, load_options=load_options)
+        if load_options.load_optimizer:
+            load_optim_checkpoint(
+                state.model, state.optimizers[0], load_path, sharded_checkpoint=load_options.sharded_checkpoint
```
Comment (bigning) [inline composer/checkpoint/load.py:125]: i remember with V2, model and optimizer state are saved in different folder ? Should we use different load path here ? I don't find the code that append the model and optimizer folder name to the load_path inside the `load_model_checkpoint` and  `load_optim_checkpoint`

File: composer/checkpoint/load.py [hunk 2]
```diff
@@ -0,0 +1,519 @@
...
+            )
+
+    if load_options.load_resumption_state:
+        load_resumption_checkpoint(state)
+
```
Comment (bigning) [inline composer/checkpoint/load.py:130]: do we need to also load the composer metadata saved by [this](https://github.com/mosaicml/composer/blob/dev/composer/checkpoint/save.py#L134) ? Or are they loaded in `load_resumption_checkpoint` ?

File: composer/checkpoint/load.py [hunk 3]
```diff
@@ -0,0 +1,519 @@
...
+    if load_options is None:
+        load_options = CheckpointLoadOptions()
+    elif isinstance(load_options, dict):
+        load_options = CheckpointLoadOptions(**load_options)
+    if load_options.include_keys is not None and load_options.ignore_keys is not None:
+        raise ValueError('Only one of include_keys or ignore_keys can be specified.')
+
+    if load_options.include_keys is not None or load_options.ignore_keys is not None:
+        load_options.strict = True
+    #assert seed == 1, f"{sharded_checkpoint=}, sharded_model={_is_model_fsdp(model)}, {shard_as_needed_during_load=}"
```
[the remark this answers is not in the document]
Comment (eracah) [reply composer/checkpoint/load.py]: fixed

File: composer/checkpoint/load.py [hunk 4]
```diff
@@ -0,0 +1,519 @@
...
+def download_and_load_sharded_state_dict(
+    load_path: str, device_mesh: Optional[DeviceMesh], state_dict: dict, load_planner: Optional[LoadPlanner] = None
+):
+    load_path_is_remote = is_uri(load_path)
+    download_dir_context = tempfile.TemporaryDirectory if load_path_is_remote else contextlib.nullcontext
+    with download_dir_context() as download_dir:
+        if load_path_is_remote:
+            storage_reader = DistCPObjectStoreReader(
+                source_path=load_path,
+                destination_path=download_dir,
```
[the remark this answers is not in the document]
Comment (eracah) [reply composer/checkpoint/load.py:467]: oh good call! fixed

File: composer/checkpoint/load.py [hunk 5]
```diff
@@ -0,0 +1,519 @@
...
+
+    if load_options.load_resumption_state:
+        load_resumption_checkpoint(state)
+
+
+def load_model_checkpoint(
+    model: ComposerModel,
+    load_path: str,
+    load_options: Optional[Union[CheckpointLoadOptions, Dict]] = None,
+    optimizer: Optional[torch.optim.Optimizer] = None,
```
[the remark this answers is not in the document]
Comment (eracah) [reply composer/checkpoint/load.py:136]: good catch fixed!

File: composer/checkpoint/load.py [hunk 6]
```diff
@@ -0,0 +1,519 @@
...
+    load_model: bool = True
+    load_optimizer: bool = False
+    load_resumption_state: bool = False  # dataset_state, timestamp,
+    # Specific key-level loading configs
+    # e.g.'model.layers.transformer.0.w1.weight'
+    include_keys: Optional[Union[str, Sequence[str]]] = None
+    ignore_keys: Optional[Union[str, Sequence[str]]] = None
+    sharded_checkpoint: bool = False  # TODO: Auto-detect sharded
+    shard_as_needed_during_load: bool = False
+    strict: bool = False
```
[the remark this answers is not in the document]
Comment (eracah) [reply composer/checkpoint/load.py]: oh yeah sure!

File: composer/checkpoint/load.py [hunk 7]
```diff
@@ -0,0 +1,519 @@
...
+    """
+    if load_options is None:
+        load_options = CheckpointLoadOptions()
+    elif isinstance(load_options, dict):
+        load_options = CheckpointLoadOptions(**load_options)
+    if load_options.include_keys is not None and load_options.ignore_keys is not None:
+        raise ValueError('Only one of include_keys or ignore_keys can be specified.')
+
+    if load_options.include_keys is not None or load_options.ignore_keys is not None:
+        load_options.strict = True
```
[the remark this answers is not in the document]
Comment (eracah) [reply composer/checkpoint/load.py:151]: good catch!

File: composer/checkpoint/load.py [hunk 8]
```diff
@@ -0,0 +1,519 @@
...
+            'Model is sharded but checkpoint is not sharded. Please pass in a model not wrapped with FSDP.'
+        )
+    load_path_is_remote = is_uri(load_path)
+    download_dir_context = tempfile.TemporaryDirectory if load_path_is_remote else contextlib.nullcontext
+    with download_dir_context() as download_dir:
+        file_path = load_path
+        if load_path_is_remote:
+            filename = Path(load_path).name
+            file_path = os.path.join(download_dir, filename)
+            download_monolithic_checkpoint(load_path, download_dir)
```
[the remark this answers is not in the document]
Comment (eracah) [reply composer/checkpoint/load.py:245]: no, see [here](https://github.com/mosaicml/composer/pull/3452/files#diff-892cea885382a643f30b5c44c013c4165065b19c42779e36951af796054cc8dbR228-R229) we return if non-global rank 0 and then [here](https://github.com/mosaicml/composer/pull/3452/files#diff-892cea885382a643f30b5c44c013c4165065b19c42779e36951af796054cc8dbR165-R170) we make sure we wrap with `sync_module_states=True` to ensure broadcasting of the weights from global rank 0

File: composer/checkpoint/load.py (hunk 2 above)
Comment (eracah) [reply composer/checkpoint/load.py:130]: we don't need to load the metadata I don't think. it's just used for someone to inspect a previous run

File: composer/checkpoint/load.py (hunk 1 above)
Comment (eracah) [reply composer/checkpoint/load.py]: oh good catch. fixed!

**Review by bigning — APPROVED** [review]

**Comment by bigning** [comment]
btw, please update the PR description with the design

**Pull request merged**